pad_server/src/controller.c

```diff
@@ -102,72 +101,88 @@ static ssize_t controller_recv(controller_t *controller, void *buf, size_t n) {
  * TODO: docs
  */
 void *controller_run(void *arg) {
-
     controller_args_t *args = (controller_args_t *)(arg);
     controller_t controller;
     int err;
 
-    /* Connect to the controller */
-    err = controller_init(&controller, args->port);
-    if (err) {
-        fprintf(stderr, "Could not initialize controller with error: %s\n", strerror(err));
-        thread_return(err);
-    }
-    pthread_cleanup_push(controller_cleanup, &controller);
+    fprintf(stderr, "Waiting for controller...\n");
 
-    err = controller_accept(&controller);
-    if (err) {
-        fprintf(stderr, "Could not connect to controller with error: %s\n", strerror(err));
-        thread_return(err);
-    }
-    printf("Controller connected!\n");
-
-    /* Receive messages */
+    pthread_cleanup_push(controller_cleanup, &controller);
 
     for (;;) {
-
-        /* Get the message header to determine what to handle */
-        header_p hdr;
-        ssize_t bread;
-        bread = controller_recv(&controller, &hdr, sizeof(hdr));
-
-        if (bread == -1) {
-            fprintf(stderr, "Error reading message header: %s\n", strerror(errno));
-            thread_return(EXIT_FAILURE);
-        } else if (bread == 0) {
-            fprintf(stderr, "Control box disconnected.\n");
-            thread_return(EXIT_FAILURE);
+        /* Initialize the controller (creates a new socket) */
+        err = controller_init(&controller, args->port);
+        if (err) {
+            fprintf(stderr, "Could not initialize controller with error: %s\n", strerror(err));
+            sleep(1);
```

Comment:
Remove all the sleep calls and just let it spin; adding a second of sleep time during disconnect adds latency to our system which might compromise control of the hybrid.